plugin/metadata/metadata.go

```diff
@@ -31,9 +31,16 @@ func (m *Metadata) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Ms
 	if plugin.Zones(m.Zones).Matches(state.Name()) != "" {
 		// Go through all Providers and collect metadata.
 		for _, provider := range m.Providers {
-			for _, varName := range provider.MetadataVarNames() {
-				if val, ok := provider.Metadata(ctx, state, varName); ok {
-					md.SetValue(varName, val)
+			for _, v := range provider.VarStrings() {
+				val := provider.String(ctx, state, v)
+				if val != "" {
+					md.SetString(v, val)
+				}
+			}
+			for _, v := range provider.VarInts() {
+				val := provider.Int(ctx, state, v)
+				if val != -1 {
+					md.SetInt(v, val)
```

Comment:
See other comment.

My fear is also here : for the plugins that need data, but do not care the type, it will have to loop over each type to get the right info ...

Reply:
Although here we could implement a convenient function ... but that will come as the same as defining a enumType for the variable.